### PR TITLE
UBERF-4321 Refactor mergeQueries

### PR DESCRIPTION
**Pull request by aonnikov**
# Contribution checklist

## Brief description
mergeQueries implementation was very naive and did not handle properly many cases. In this PR I modified it a bit to be more stable. we still need to support more predicates, though

## Checklist

* [ ] - UI test added to added/changed functionality?
* [ ] - Screenshot is added to PR if applicable ?
* [ ] - Does the code work? Check whether function and logic are correct.
* [ ] - Does Changelog.md is updated with changes?
* [ ] - Does the translations are up to date?
* [ ] - Does it well tested?
* [ ] - Tested for Chrome.
* [ ] - Tested for Safari.
* [ ] - Go through the changed code looking for typos, TODOs, commented LOCs, debugging pieces of code, etc.
* [ ] - Rebase your branch onto master and upstream branch
* [ ] - Is there any redundant or duplicate code?
* [ ] - Are required links are linked to PR?
* [ ] - Does new code is well documented ?

## Related issues

A list of closed updated issues

### CODE REVIEW

File: packages/core/src/__tests__/utils.test.ts [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    const q1 = { name: 'john' }
+    const q2 = { age: { $gt: 42 } }
+    const res = { name: 'john', age: { $gt: 42 } }
+
+    expect(mergeQueries(q1, q2)).toEqual(res)
+    expect(mergeQueries(q2, q1)).toEqual(res)
+  })
+
+  it('merges equal field values', () => {
+    expect(mergeQueries({ value: 42 }, { value: 42 })).toEqual({ value: 42 })
```
Comment (haiodo) [inline packages/core/src/__tests__/utils.test.ts:38]: Why two times?

File: packages/core/src/__tests__/utils.test.ts [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    const res = { value: { $in: [] } }
+    expect(mergeQueries(q1, q2)).toEqual(res)
+    expect(mergeQueries(q2, q1)).toEqual(res)
+  })
+
+  it('merges predicate with predicate', () => {
+    expect(mergeQueries({ age: { $in: [41, 42] } }, { age: { $in: [42, 43] } })).toEqual({ age: { $in: [42] } })
+    expect(mergeQueries({ age: { $in: [42, 43] } }, { age: { $in: [41, 42] } })).toEqual({ age: { $in: [42] } })
+
+    expect(mergeQueries({ age: { $nin: [42] } }, { age: { $nin: [42] } })).toEqual({ age: { $nin: [42] } })
```
Comment (haiodo) [inline packages/core/src/__tests__/utils.test.ts:53]: Duplicate line

File: packages/core/src/__tests__/utils.test.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    expect(mergeQueries({ age: { $gt: 40 } }, { age: { $gt: 42 } })).toEqual({ age: { $gt: 42 } })
+    expect(mergeQueries({ age: { $gt: 42 } }, { age: { $gt: 40 } })).toEqual({ age: { $gt: 42 } })
+
+    expect(mergeQueries({ age: { $lte: 45 } }, { age: { $lte: 42 } })).toEqual({ age: { $lte: 42 } })
+    expect(mergeQueries({ age: { $lte: 42 } }, { age: { $lte: 45 } })).toEqual({ age: { $lte: 42 } })
+
+    expect(mergeQueries({ age: { $gte: 40 } }, { age: { $gte: 42 } })).toEqual({ age: { $gte: 42 } })
+    expect(mergeQueries({ age: { $gte: 42 } }, { age: { $gte: 40 } })).toEqual({ age: { $gte: 42 } })
+
+    expect(mergeQueries({ age: { $ne: 42 } }, { age: { $ne: 42 } })).toEqual({ age: { $nin: [42] } })
```
Comment (haiodo) [inline packages/core/src/__tests__/utils.test.ts]: We need to be aware performance are same in this case.

File: packages/core/src/__tests__/utils.test.ts [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+  })
+
+  it('$nin merge', () => {
+    expect(mergeQueries({ value: { $nin: [] } }, { value: { $nin: [] } })).toEqual({ value: { $nin: [] } })
+    expect(mergeQueries({ value: { $nin: [42] } }, { value: { $nin: [42] } })).toEqual({ value: { $nin: [42] } })
+  })
+
+  it('$in and $nin merge', () => {
+    expect(mergeQueries({ value: { $in: [1, 2] } }, { value: { $nin: [1] } })).toEqual({
+      value: { $in: [1, 2], $nin: [1] }
```
Comment (haiodo) [inline packages/core/src/__tests__/utils.test.ts]: Should it be in [2] only?

Comment (aonnikov) [reply packages/core/src/__tests__/utils.test.ts]: yeah ... need to merge this together

File: packages/core/src/__tests__/utils.test.ts (hunk 1 above)
Comment (aonnikov) [reply packages/core/src/__tests__/utils.test.ts:38]: mistake, removed

File: packages/core/src/__tests__/utils.test.ts (hunk 2 above)
Comment (aonnikov) [reply packages/core/src/__tests__/utils.test.ts:53]: mistake, removed

File: packages/core/src/__tests__/utils.test.ts (hunk 3 above)
Comment (aonnikov) [reply packages/core/src/__tests__/utils.test.ts]: I modified this to preserve original operator when possible.

File: packages/core/src/__tests__/utils.test.ts (hunk 4 above)
Comment (aonnikov) [reply packages/core/src/__tests__/utils.test.ts]: Yes, fixed this and more bugs

**Review by haiodo — APPROVED** [review]

**Pull request merged**